mayastor/src/bdev/nexus/nexus_share.rs

```diff
@@ -12,33 +12,39 @@ use crate::{
             DestroyCryptoBdev,
             Error,
             Nexus,
-            ShareNexus,
+            ShareIscsiNexus,
+            ShareNbdNexus,
         },
-        nexus_nbd::Disk,
+        nexus_iscsi::NexusIscsiTarget,
+        nexus_nbd::NbdDisk,
     },
     core::Bdev,
     ffihelper::{cb_arg, done_errno_cb, errno_result_from_i32, ErrnoResult},
 };
 
+use rpc::mayastor::{
+    ShareProtocolNexus,
+};
+
 /// we are using the multi buffer encryption implementation using CBC as the
 /// algorithm
 const CRYPTO_FLAVOUR: &str = "crypto_aesni_mb";
 
 impl Nexus {
-    /// Publish the nexus to system using nbd device and return the path to
-    /// nbd device.
     pub async fn share(
         &mut self,
+        share_protocol: ShareProtocolNexus,
         key: Option<String>,
     ) -> Result<String, Error> {
-        if self.nbd_disk.is_some() {
+
+        assert_eq!(self.share_handle, None);
+
+        if self.share_protocol.is_some() {
```

Comment:
This is tricky! First off, if we ditch the Option we can write this as:

```rust
if self.share_protocol != ShareProtocol::None {
// we should already be shared -- as CSI is idempotent chances are we get called for some odd 
// reason. Validate indeed -- that we are shared by walking the target. If so,  simply return Ok()
// if we are not shared but the variant says we should be, try to correct the state if the variants are the 
// same. If they are not, return an error making sure we return the right error.
}
```

Reply:
Resolved in https://github.com/openebs/Mayastor/pull/125/commits/d2abbaa1d04637148b05dac687b3b713b529794b